src/rules/preferReadonlyRule.ts

```diff
@@ -0,0 +1,158 @@
+/**
+ * @license
+ * Copyright 2013 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as utils from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static metadata: Lint.IRuleMetadata = {
+        description: "Requires that private variables are marked as `readonly` if they're never modified outside of the constructor.",
+        descriptionDetails: Lint.Utils.dedent`
+            If a variable is only assigned to in the constructor, it should be declared as \`readonly\`.
+        `,
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        ruleName: "prefer-readonly",
+        type: "maintainability",
+        typescriptOnly: true,
+    };
+
+    public static FAILURE_STRING_FACTORY(identifier: string) {
+        return `Private member variable '${identifier}' is never reassigned; mark it as 'readonly'.`;
+    }
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        const preferReadonlyWalker = new PreferReadonlyWalker(sourceFile, this.ruleName, undefined);
+        return this.applyWithWalker(preferReadonlyWalker);
+    }
+}
+
+class ClassScope {
+    public declaredMembers = new Map<string, ts.Identifier>();
+    public modifiedMembers = new Set<string>();
+    public inConstructor = false;
+}
+
+class PreferReadonlyWalker extends Lint.AbstractWalker<void> {
+    private scope: ClassScope;
+
+    public walk(sourceFile: ts.SourceFile) {
+        if (sourceFile.isDeclarationFile) {
+            return;
+        }
+
+        ts.forEachChild(sourceFile, this.visitNode);
+    }
+
+    private readonly visitNode = (node: ts.Node) => {
+        switch (node.kind) {
+            case ts.SyntaxKind.ClassDeclaration:
+                this.handleClassDeclaration(node as ts.ClassDeclaration);
+                break;
+
+            case ts.SyntaxKind.Constructor:
+                this.handleConstructor(node as ts.ConstructorDeclaration);
+                break;
+
+            case ts.SyntaxKind.PropertyDeclaration:
+                this.handlePropertyDeclaration(node as ts.PropertyDeclaration);
+                break;
+
+            case ts.SyntaxKind.PropertyAccessExpression:
```

Comment:
There's no other choice. You need to know if the property is assigned on an object of a (sub)type of the current class. Consider the following code:

```ts
class A {
    private foo;
    private bar;
    private baz;

    public copyTo(other: A) {
        other.foo = this.foo; // this one needs to be recognized as assignment
    }

    public copy(other: {[key: string]: any} {
        // both should not be handled as assignments to `A#bar` or `A#baz`
        other.bar = this.bar;
        other.baz = this.baz;
    }
}
```